File: PDOO/Napakalaki-Java/src/NapakalakiGame/NumericBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Cervi
 */
public class NumericBadConsequence extends BadConsequence{
    
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    
    public NumericBadConsequence(String text, int levels, int nVisible, int nHidden){
        super(text,levels);
        this.nVisibleTreasures=nVisible;
        this.nHiddenTreasures=nHidden;
    }
    
    public int getNHiddenTreasures(){
        return this.nHiddenTreasures;
    }
    
    public int getNVisibleTreasures(){
        return this.nVisibleTreasures;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        if(nVisibleTreasures>0)
            nVisibleTreasures--;
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        if(nHiddenTreasures>0)
            nHiddenTreasures--;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        int nVret, nHret;
        
        int nVis = v.size();
        int nHid = h.size();
        
        if(nVis<nVisibleTreasures)
            nVret=nVis;
        else
            nVret=nVisibleTreasures;
        
        if(nHid<nHiddenTreasures)
            nHret=nHid;
        else
            nHret=nHiddenTreasures;
        
        BadConsequence adjustedBC=new NumericBadConsequence(text,0, nVret, nHret);
        System.out.println("MAL ROLLO AJUSTADO: "+adjustedBC.toString());
        return adjustedBC;
    }
    
    @Override
    public boolean isEmpty(){
        if(nVisibleTreasures == 0 && nHiddenTreasures==0)
            return true;
        else
            return false;
    }
    
    @Override
    public String toString(){
        String description = "--- Mal Rollo  --- \n" + "Texto: " + text + "\n";
            description += "Pérdida de niveles: " + levels + "\n";
            description += "Pérdida de tesoros visibles: " + nVisibleTreasures + "\n";
            description += "Pérdida de tesoros ocultos: " + nHiddenTreasures + "\n";
    
        return description;
    }
}

File: PDOO/Napakalaki-Ruby/lib/NumericBadConsequence.rb
#encoding: utf-8

require 'BadConsequence.rb'
require 'Treasure.rb'

module NapakalakiGame
    
  class NumericBadConsequence < BadConsequence
      
    def initialize(text,levels,nVisible,nHidden)
      super(text,levels)
      @nHiddenTreasures = nHidden
      @nVisibleTreasures = nVisible
    end
    
    def isEmpty()
       if(@nVisibleTreasures == 0 && @nHiddenTreasures == 0)
         return true
       end
    return false
    end
    
    def getNVisibleTreasures()
      return @nVisibleTreasures
    end
    
    def getNHiddenTreasures()
      return @nHiddenTreasures
    end
    
    def substractVisibleTreasures(t)
      @nVisibleTreasures -= 1
    end
    
    def substractHiddenTreasures(t)
      @nHiddenTreasures -= 1
    end
    
    def adjustToFitTreasuresLists(visible,hidden)
      if(@nVisibleTreasures <= visible.length)
        vRet = @nVisibleTreasures
      else
        vRet = visible.length
      end
      if(@nHiddenTreasures <= hidden.length)
        hRet = @nHiddenTreasures
      else
        hRet = hidden.length
      end
      ret = NumericBadConsequence.new(@text,0, vRet, hRet)
      return ret
    end
    
    def to_s()
      return @text.to_s() +"\n\t\tNiveles:"+@levels.to_s()+"\n\t\tnHidden:"+@nHiddenTreasures.to_s()+"\n\t\tnVisible:"+@nVisibleTreasures.to_s()
    end
    
  end
end
